state/sqlserver/migration.go

```diff
@@ -275,32 +274,61 @@ func (m *migration) ensureUpsertStoredProcedureExists(db *sql.DB, mr migrationRe
 		CREATE PROCEDURE %s (
 			@Key 			%s,
 			@Data 			NVARCHAR(MAX),
-			@RowVersion 	BINARY(8))
+			@RowVersion 	BINARY(8),
+			@FirstWrite		BIT)
 		AS
 			IF (@RowVersion IS NOT NULL)
 			BEGIN
 				UPDATE [%s]
 				SET [Data]=@Data, UpdateDate=GETDATE()
 				WHERE [Key]=@Key AND RowVersion = @RowVersion
-
 				RETURN
 			END
 
-			BEGIN TRY
-				INSERT INTO [%s] ([Key], [Data]) VALUES (@Key, @Data);
-			END TRY
-
-			BEGIN CATCH
-				IF ERROR_NUMBER() IN (2601, 2627)
-				UPDATE [%s]
-				SET [Data]=@Data, UpdateDate=GETDATE()
-				WHERE [Key]=@Key AND RowVersion = ISNULL(@RowVersion, RowVersion)
-			END CATCH`,
+			IF (@FirstWrite=1)
+			BEGIN
+				BEGIN TRANSACTION;
+				IF EXISTS (SELECT * FROM [%s] WHERE [KEY]=@KEY)
+				BEGIN
+					THROW 2714, ''FIRST-WRITE: COMPETING RECORD ALREADY WRITTEN.'', 1
```

Comment:
It's unusual that you're returning an error 2714 which is reserved for objects (i.e. tables, databases) which rows are not generally considered in SQL. 

I assume SQL would return 2601/2627 since that is also what this proc checks for to switch from insert to update, so either of those are probably more idiomatic errors to return. Given that this is a stored proc, it may also be more useful to `RAISERROR` with a specific error message that describes the Dapr-enforced first-write wins constraint here instead.

Reply:
RAISERROR did not work at all the way the stored procedure is created here. It did work in a SQL editor I was using but over 2 hours I could not work around that. THROW works fine -- I can change to any error code that is desirable.

Reply:
I tried RAISEERROR again but by design this doesn't exit the SQL execution flow. SQL Server engineers recommend using THROW.